assets/js/components/ProjectEditorTextField.js

```diff
@@ -0,0 +1,184 @@
+              showError(data)
+              reject(statusCode)
+            } else {
+              resolve(statusCode)
+            }
+          },
+          error: function (error) {
+            if (error.status === 422) {
+              error.message = error.responseText
+              showError(error)
+              reject(error.status)
+            } else if (error.status === 401) {
+              reject(error.status)
+            }
+          }
+        })
+      })
+    } else if (newText === '') {
+      return this.customTranslationApi.deleteCustomTranslation(
+        this.programId,
+        languageSelected,
+        deleteCustomTranslationSuccess,
+        showError
+      )
+    } else {
+      return this.customTranslationApi.saveCustomTranslation(
+        this.programId,
+        newText,
+        languageSelected,
+        saveCustomTranslationSuccess,
+        showError
+      )
+    }
+  }
+
+  function showError (error) {
+    self.editText.addClass('danger')
+    self.editTextError.show()
+    self.editTextError.text(error.message)
+  }
+
+  function deleteCustomTranslationSuccess (language) {
+    self.snackbar.show(self.translationDeleted, language)
+    self.hideError()
+  }
+
+  function saveCustomTranslationSuccess (language) {
+    self.snackbar.show(self.translationSaved, language)
+    self.hideError()
+  }
+
+  function getCustomTranslationSuccess (data) {
+    self.textLoadingSpinner.hide()
+    self.editText.val(data)
+    self.lastSavedText = data
+    self.editText.removeAttr('disabled')
+  }
+
+  function getCustomTranslationError () {
+    self.textLoadingSpinner.hide()
+    self.editText.val('')
+    self.lastSavedText = ''
+    self.editText.removeAttr('disabled')
+  }
+
+  this.getNewText = (language) => {
+    if (language === 'default') {
+      this.editText.removeAttr('disabled')
+      this.editText.val(this.initialText)
+      this.lastSavedText = this.initialText
+      return
+    } else if (this.isDisabled(language)) {
+      this.lastSavedText = this.initialText
+      this.editText.val(this.initialText)
+      this.editText.attr('disabled', '')
+      return
+    }
+
+    this.editText.attr('disabled', '')
+    this.editText.val('')
+    this.textLoadingSpinner.show()
+
+    this.customTranslationApi.getCustomTranslation(
+      this.programId,
+      language,
+      getCustomTranslationSuccess,
+      getCustomTranslationError
+    )
+  }
+
+  function setLastSavedText (data) {
+    self.lastSavedText = data
+  }
+
+  function setLastSavedTextError () {
+    self.lastSavedText = ''
+  }
+
+  this.keepChanges = (language) => {
+    if (this.isDisabled(language)) {
+      this.lastSavedText = this.initialText
+      this.editText.val(this.initialText)
+      this.editText.attr('disabled', '')
+    } else if (language === 'default') {
+      this.editText.removeAttr('disabled')
+      this.lastSavedText = this.initialText
+    } else {
+      this.customTranslationApi.getCustomTranslation(
```

Comment:
should we show a loading spinner here?

Reply:
I think it is fine if this loads in the background. They would have to save it pretty fast for this to cause a problem

Reply:
ok